ethereum/eip712/eip712_fuzzer_test.go

```diff
@@ -0,0 +1,140 @@
+package eip712_test
+
+import (
+	"fmt"
+	"strings"
+
+	rand "github.com/tendermint/tendermint/libs/rand"
+
+	"github.com/evmos/evmos/v11/ethereum/eip712"
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+)
+
+// TestRandomPayloadFlattening generates many random payloads with different JSON values to ensure
+// that Flattening works across all inputs.
+// Note that this is a fuzz test, although it doesn't use Go's Fuzz testing suite, since there are
+// variable input sizes, types, and fields. While it may be possible to translate a single input into
+// a JSON object, it would require difficult parsing, and ultimately approximates our randomized unit
+// tests as they are.
+func (suite *EIP712TestSuite) TestRandomPayloadFlattening() {
+	// Re-seed rand generator
+	rand.Seed(rand.Int64())
+
+	numTestObjects := 15
+	for i := 0; i < numTestObjects; i++ {
+		suite.Run(fmt.Sprintf("Flatten%d", i), func() {
+			payload := suite.generateRandomPayload(i)
+
+			flattened, numMessages, err := eip712.FlattenPayloadMessages(payload)
+
+			suite.Require().NoError(err)
+			suite.Require().Equal(numMessages, i)
+
+			suite.verifyPayloadAgainstFlattened(payload, flattened)
+		})
+	}
+}
+
+// generateRandomPayload creates a random payload of the desired format, with random sub-objects.
+func (suite *EIP712TestSuite) generateRandomPayload(numMessages int) gjson.Result {
+	payload := suite.createRandomJSONObject().Raw
+	msgs := make([]gjson.Result, numMessages)
+
+	for i := 0; i < numMessages; i++ {
+		m := suite.createRandomJSONObject()
+		msgs[i] = m
+	}
+
+	payload, err := sjson.Set(payload, "msgs", msgs)
+	suite.Require().NoError(err)
+
+	return gjson.Parse(payload)
+}
+
+// createRandomJSONObject creates a JSON object with random fields.
+func (suite *EIP712TestSuite) createRandomJSONObject() gjson.Result {
+	var err error
+	payloadRaw := ""
+
+	numFields := suite.randomInRange(0, 16)
+	for i := 0; i < numFields; i++ {
+		key := suite.generateRandomString(12, 36)
+
+		randField := suite.createRandomJSONField(i, 0)
+		payloadRaw, err = sjson.Set(payloadRaw, key, randField)
+		suite.Require().NoError(err)
+	}
+
+	return gjson.Parse(payloadRaw)
+}
+
+// createRandomJSONField creates a random field with a random JSON type, with the possibility of
+// nested fields up to depth.
+func (suite *EIP712TestSuite) createRandomJSONField(t int, depth int) interface{} {
```

Comment:
Just for context, I originally decided against using native Golang fuzzing since we want to test JSON objects of indeterminate size with a fixed number of fuzzing inputs. 

If we were to use the native fuzzing library, we could do so in the following way: take a byte stream as input, then hash it byte-by-byte (pre-pended with a salt), then for each hash, we convert it into a JSON object with random values, in effect using it as a seed.

I didn't see the value in going through this trouble because it ultimately gives us the same result—it becomes a seeded random number generator, which we use to create test values. As such, we wouldn't get much utility out of the fuzzing library.